cmd/gpu_nfdhook/labeler.go

```diff
@@ -0,0 +1,209 @@
+
+type labeler struct {
+	sysfsDir      string
+	devfsDir      string
+	debugfsDRIDir string
+
+	gpuDeviceReg     *regexp.Regexp
+	controlDeviceReg *regexp.Regexp
+	labels           labelMap
+}
+
+func newLabeler(sysfsDir, devfsDir, debugfsDRIDir string) *labeler {
+	return &labeler{
+		sysfsDir:         sysfsDir,
+		devfsDir:         devfsDir,
+		debugfsDRIDir:    debugfsDRIDir,
+		gpuDeviceReg:     regexp.MustCompile(gpuDeviceRE),
+		controlDeviceReg: regexp.MustCompile(controlDeviceRE),
+		labels:           labelMap{},
+	}
+}
+
+func (l *labeler) scan() ([]string, error) {
+	files, err := ioutil.ReadDir(l.sysfsDir)
+	gpuNameList := []string{}
+
+	if err != nil {
+		return gpuNameList, errors.Wrap(err, "Can't read sysfs folder")
+	}
+
+	for _, f := range files {
+		if !l.gpuDeviceReg.MatchString(f.Name()) {
+			klog.V(4).Info("Not compatible device", f.Name())
+			continue
+		}
+
+		dat, err := ioutil.ReadFile(path.Join(l.sysfsDir, f.Name(), "device/vendor"))
+		if err != nil {
+			klog.Warning("Skipping. Can't read vendor file: ", err)
+			continue
+		}
+
+		if strings.TrimSpace(string(dat)) != vendorString {
+			klog.V(4).Info("Non-Intel GPU", f.Name())
+			continue
+		}
+
+		drmFiles, err := ioutil.ReadDir(path.Join(l.sysfsDir, f.Name(), "device/drm"))
+		if err != nil {
+			return gpuNameList, errors.Wrap(err, "Can't read device folder")
+		}
+
+		for _, drmFile := range drmFiles {
+			if l.controlDeviceReg.MatchString(drmFile.Name()) {
+				//Skipping possible drm control node
+				continue
+			}
+			devPath := path.Join(l.devfsDir, drmFile.Name())
+			if _, err := os.Stat(devPath); err != nil {
+				continue
+			}
+
+			gpuNameList = append(gpuNameList, f.Name())
+			break
+		}
+	}
+
+	return gpuNameList, nil
+}
+
+// getMemoryValues reads the GPU memory amount from the system.
+func (l *labeler) getMemoryAmount( /*cardNum*/ string) uint64 {
+	// reading GPU local memory amount is not yet available in the driver,
+	// so just return the environment variable value
+	envValue := os.Getenv(memoryOverrideEnv)
+	if envValue != "" {
+		val, err := strconv.ParseUint(envValue, 10, 64)
+		if err == nil {
+			return val
+		}
+	}
+	return 0
+}
+
+// addNumericLabel creates a new label if one doesn't exist. Else the new value is added to the previous value.
+func (lm labelMap) addNumericLabel(labelName string, valueToAdd int64) {
+	value := int64(0)
+	if numstr, ok := lm[labelName]; ok {
+		_, _ = fmt.Sscanf(numstr, "%d", &value)
+	}
+	value += valueToAdd
+	lm[labelName] = strconv.FormatInt(value, 10)
+}
+
+// createCapabilityLabels creates labels from the gpu capability file under debugfs.
+func (l *labeler) createCapabilityLabels(cardNum string) {
+	// try to read the capabilities from the i915_capabilities file
+	file, err := os.Open(filepath.Join(l.debugfsDRIDir, cardNum, "i915_capabilities"))
+	if err != nil {
+		klog.V(3).Infof("Couldn't open file:%s\n", err.Error()) // debugfs is not stable, there is no need to spam with error level prints
```

Comment:
klog adds new lines always, no need to put it after the message manually.

Reply:
cosmetics fixed